cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -349,6 +370,47 @@ public Void call() throws Exception {
     });
   }
 
+  private ProgramState getProgramState(boolean success, String failureInfo) {
+    if (stopRequested) {
+      // Program explicitly stopped, return KILLED state
+      return new ProgramState(ProgramStatus.KILLED, null);
+    }
+    if (!success) {
+      // Program is unsuccessful, return FAILED state
+      return new ProgramState(ProgramStatus.FAILED, failureInfo);
+    }
+    // Program is successfully completed, return COMPLETE state
+    return new ProgramState(ProgramStatus.COMPLETED, null);
+  }
+
+  /**
+   * Calls the destroy method of {@link ProgramLifecycle}.
+   */
+  private void destroy(final boolean succeeded, final BasicSparkClientContext context,
```

Comment:
I kept it as a separate method because `onFinish` method throws an exception, while contract of the `destroy` is not to throw an exception but log as warning in case if any.